aip/0153.md

```diff
@@ -0,0 +1,164 @@
+---
+aip:
+  id: 153
+  state: reviewing
+  created: 2019-12-16
+permalink: /153
+redirect_from:
+  - /0153
+---
+
+# Import and export
+
+Many users want to be able to load data into an API, or get their existing data
+out of an API. This is particularly important for enterprise users who are
+concerned about vendor lock.
+
+## Guidance
+
+APIs **may** support import and export when they desire to do so. Additionally,
+import and export methods **may** create multiple new resources, or they
+**may** populate data into a single resource.
+
+### Importing multiple resources
+
+Services **should** implement a common pattern for importing multiple resources
+into an API:
+
+```proto
+rpc ImportBooks(ImportBooksRequest) returns (google.longrunning.Operation) {
+  option (google.api.http) = {
+    post: "/v1/{parent=publishers/*}/books:import"
+    body: "*"
+  };
+  option (google.longrunning.operation_info) = {
+    response_type: "ImportBooksResponse"
+    metadata_type: "ImportBooksMetadata"
+  };
+}
+```
+
+- The method **must** return a long-running operation (see [AIP-151][]) unless
+  the service can guarantee that it will _never_ need more than a few seconds
+  to complete.
+- The HTTP verb **must** be `POST`, and the `body` **must** be `"*"`.
+- A `parent` field **should** be included as part of the URI.
+  - If importing into or exporting from multiple resources is required, the API
+    **should** keep the `parent` field and allow the user to use the `-`
+    character to indicate multiple parents (see [AIP-159][]).
+  - On import, if the user provides a specific parent, the API **must** reject
+    any imported resources that would be added to a different parent.
+- The URI suffix **should** be `:import` or `:export`.
+
+### Importing data into a resource
```

Comment:
This section seems half "only about import" (only an import example, only import in the title) but there's no equivalent export section, and the last bullet point talks about export as well. We should decide one way or the other. (There's a sort of reference to this right near the end, but it feels a little tacked on.)

Reply:
I struggled with this throughout the document. It feels really redundant to add export examples to everything, but at the same time, I want it to be clear to readers that we are covering both.

Reply:
Yup - and it makes it easier to give any import-specific advice in an import section, and export-specific advice in an export section.

Reply:
Right, but I do not really want each of these import sections to have a corresponding export section, which I am concerned would make the document very long. Let me see what I can do to tighten this up.